data_layout.go

```diff
@@ -0,0 +1,421 @@
+	floatValsLen := len(floatVals)
+	size := getUInt64(strings.TrimPrefix(floatVals[0], "f"))
+	abi := getUInt64(floatVals[1])
+	pref := abi
+	if floatValsLen == 3 {
+		pref = getUInt64(floatVals[2])
+	}
+	dl.floatSizeAlign[size] = NewFloatingPointSizeAlignment(size, abi, pref)
+}
+
+func addAggAlignFromString(spec string, dl *DataLayout) {
+	aggVals := strings.Split(spec, ":")
+	aggValsLen := len(aggVals)
+	abi := getUInt64(aggVals[1])
+	pref := abi
+	if aggValsLen == 3 {
+		pref = getUInt64(aggVals[2])
+	}
+	dl.aggAlign = NewAggregateAlignment(abi, pref)
+}
+
+func addFuncPtrAlignFromString(spec string, dl *DataLayout) {
+	val := strings.TrimPrefix(spec, "F")
+	var typ string
+	if strings.HasPrefix(val, "i") {
+		typ = "i"
+	} else if strings.HasPrefix(val, "n") {
+		typ = "n"
+	} else {
+		panic("Invalid Function pointer alignment type, possible options are \"i\", \"n\".")
+	}
+	abi := getUInt64(strings.TrimPrefix(val, typ))
+	dl.funcPtrAlign = NewFunctionPointerAlignment(typ == "i", abi)
+}
+
+func addNativeIntBitWidths(spec string, dl *DataLayout) {
+	widthContents := strings.Split(strings.TrimPrefix(spec, "n"), ":")
+	widths := make([]uint64, len(widthContents))
+	for i, width := range widthContents {
+		widths[i] = getUInt64(width)
+	}
+	dl.nativeIntBitWidths = widths
+}
+
+func getUInt64(str string) uint64 {
+	val, err := strconv.ParseUint(str, 10, 64)
+	if err != nil {
+		panic(err.Error())
+	}
+	return val
+}
+
+func (dl *DataLayout) SetIsBigEndian(isBigEndian bool) {
+	dl.isBigEndian = isBigEndian
+}
+
+func (dl *DataLayout) SetNaturalStackAlignment(al uint64) {
+	dl.naturalStackAlignment = al
+}
+
+func (dl *DataLayout) SetProgramMemoryAddressSpace(as uint64) {
+	dl.programMemoryAddressSpace = as
+}
+
+func (dl *DataLayout) SetGlobalVarAddressSpace(as uint64) {
+	dl.globalVarAddressSpace = as
+}
+
+func (dl *DataLayout) SetAllocaAddressSpace(as uint64) {
+	dl.allocaAddressSpace = as
+}
+
+func (dl *DataLayout) AddPointerSizeAlignment(instance *PointerSizeAlignment) {
+	dl.ptrSizeAlign[instance.addressSpace] = instance
+}
+
+func (dl *DataLayout) AddIntegerSizeAlignment(instance *IntegerSizeAlignment) {
+	dl.intSizeAlign[instance.size] = instance
+}
+
+func (dl *DataLayout) AddVectorSizeAlignment(instance *VectorSizeAlignment) {
+	dl.vecSizeAlign[instance.size] = instance
+}
+
+func (dl *DataLayout) AddFloatingPointSizeAlignment(instance *FloatingPointSizeAlignment) {
+	dl.floatSizeAlign[instance.size] = instance
+}
+
+func (dl *DataLayout) SetAggregateAlignment(instance *AggregateAlignment) {
+	dl.aggAlign = instance
+}
+
+func (dl *DataLayout) SetFunctionPointerAlignment(instance *FunctionPointerAlignment) {
+	dl.funcPtrAlign = instance
+}
+
+func (dl *DataLayout) SetManglingStyle(style ManglingStyle) {
+	dl.mangling = style
+}
+
+func (dl *DataLayout) SetNativeIntBitWidths(bitWidths []uint64) {
+	dl.nativeIntBitWidths = bitWidths
+}
+
+func (dl *DataLayout) GetLayoutString() string {
+	layout := ""
```

Comment:
To handle memory allocation, we could consider using `strings.Builder` to generate the string. See https://github.com/llir/llvm/blob/9e62c17a9f75d127c4010b0d242912413c9d73cc/ir/func.go#L126 for an example.